Directory API refactoring

gtristan opened the pull request:
This branch includes a series of changes to cleanup the Directory API.

Fixes #1294

gtristan committed:
storage: Rename VirtualDirectoryError -> DirectoryError

Make this name a bit more consistent as they are raised from the Directory API

Also touch up the documentation of the error.

Updated sources and tests which handle this error.

gtristan committed:
storage/directory.py: Remove _mark_changed() internal API.

This is not implemented by any directory implementation, and unused.

gtristan committed:
storage/directory.py: Removed mark_unmodified/list_modified_paths

This was much implementation plumbing just for the sake of showing
which files were modified due to integration commands being run in
the compose plugin.

I think we're better off not carrying this weight in the new API
unless there is a strong motivation to bring it back in the future.

This removes a test case which was testing this functionality, along
with the implementations in the cas/file directory backends.

gtristan committed:
storage/directory.py: Make set_deterministic_user() an internal function

There is no reason for plugin authors to have access to this API.

gtristan committed:
storage/directory.py: Added full mypy annotations to API surface

Also enhanced documentation thoroughly and demarked public/internal
sections of the API.

gtristan committed:
storage: Raise only DirectoryError from Directory backends

The API advertizes that it will raise DirectoryError for system errors,
better to raise DirectoryError than FileNotFoundError or other native python
system errors.

With the exception that we still raise ValueError for clear cut programming
errors which should not be handled by the caller but rather their code should
be fixed.

gtristan committed:
storage/_filebaseddirectory.py: Public/Private split + type annotations

Add full type annotations to all methods and properly depict private
variables and methods with double underscore, also reorganized file to
follow the order of abstract methods declared in the Directory API.

gtristan committed:
storage/directory.py: Added declaration for abstract internal method _get_underlying_directory()

gtristan committed:
storage/directory.py: Implement our own FileStat

Instead of exposing underlying python API which we are not guaranteeing
any support for, expose a limited API to represent file stats which we
can extend over time depending on what support we add over time.

gtristan committed:
storage/_filebaseddirectory.py: Implement import_files() without importing CasBasedDirectory

This also adds internal abstract method _get_underlying_path() to the Directory
so that FileBasedDirectory can perform copies and links under the hood from
other abstract Directory objects.

gtristan committed:
storage/directory.py: Remove "report_written" from Directory.import_files()

This is just unconditionally True across the board anyway.

gtristan committed:
Remove duplicate code backing `bst artifact list-contents`

There was a whole separate code path to poke into the private details
of CasBasedDirectory for the purpose of listing artifact contents.

Instead, just use the Directory API directly to implement this now
and ditch all the extra redundant codepaths.

Interestingly, this corrects the tests/frontend/artifact_list_contents.py
test which was expecting a file size of 107 bytes, which is actually
only 28 bytes long.

gtristan committed:
storage/directory.py: Move _FileType to CasBasedDirectory

This is now entirely an internal implementation detail of how
CasBasedDirectory handles it's internal IndexEntry objects.

gtristan committed:
tests/frontend/artifact_list_contents.py: Add descriptive comment

When listing the executable status of the /usr/bin/hello file we expect
a regular non-executable file, this is confusing but is due to an upstream
bug in pytest-datafiles:

    https://github.com/omarkohl/pytest-datafiles/issues/11

Save time pondering this mishap in the future by adding a comment beside
the confusing test case.

gtristan committed:
storage/directory.py: Create an internal version of Directory.import_files()

In order to hide some parameters we would rather not expose to plugin authors,
create _import_files_internal() for internal BuildStream use of this API and
expose only a simplified version.

gtristan committed:
storage/_casbaseddirectory.py: Stop importing FileBasedDirectory

This is unnecessary.

gtristan committed:
doc: Add the Directory object to the API reference index

gtristan committed:
storage/_casbaseddirectory.py: Full cleanup

  - Add full type annotations (except for remote_execution_pb2.Digest types)

  - Reorganized file into sections
    - Public API
    - Internal abstract methods
    - Internal APIs specific to CasBasedDirectory
    - Private methods

  - Removed various unused variables and functions

  - Adapted test cases in tests/internal to access some changed
    private variable names

gtristan committed:
storage/directory.py: Remove `can_link` parameter from Directory.import_files()

This is a historic artifact from when we used to hardlink files into a staging
area which is mostly rendered obsolete by allowing buildbox to handle these
details, it is ignored by the CasBasedDirectory completely, and we don't use
it anywhere with the FileBasedDirectory.

Remove the option entirely from public and private import_files() codepaths.

gtristan committed:
storage/directory.py: Make Directory.export_files() an internal function

This is not needed by plugin authors.

gtristan committed:
storage/directory.py: Revisit FileStat definition, revive FileType

After discussing this with Jürg we decided to make the file type an
enumeration again.

gtristan committed:
storage/directory.py: Switch API to use path strings instead of lists

This is a more natural and intuitive API, documented how path strings
work with "/" separators and support for "." and ".." elements.

gtristan committed:
storage/directory.py: Make get_size() an internal method

This is not needed by plugin authors and is only used internally to
calculate the size of an artifact being cached.

abderrahim commented:
I find it a little hard to review, it would be nice to split into smaller PRs for review. The following I think are uncontroversial and can easily be merged in a first batch:

- storage: Rename VirtualDirectoryError -> DirectoryError
- storage/directory.py: Remove _mark_changed() internal API.
- storage/directory.py: Removed mark_unmodified/list_modified_paths
- storage/directory.py: Make set_deterministic_user() an internal function
- storage/directory.py: Added full mypy annotations to API surface
- storage/directory.py: Added declaration for abstract internal method _get_underlying_directory()
- storage: Raise only DirectoryError from Directory backends
- storage/directory.py: Remove "report_written" from Directory.import_files()
- storage/_casbaseddirectory.py: Stop importing FileBasedDirectory
- storage/directory.py: Make get_size() an internal method


I have a couple of comments on these:
- storage/directory.py: Remove `can_link` parameter from Directory.import_files()
  Even though it is currently unused, there is potential to have a similar option that uses  the `move_files` hint  (by passing `--capture-allow-file-moves` to `buildbox-casd`). Can be done later though.
- storage/directory.py: Make Directory.export_files() an internal function
  If we want to go for https://github.com/apache/buildstream/pull/1607#issuecomment-1059019265, we'll need to keep this public. Again, this can be made public later if needed (even if it's not at 2.0).

gtristan commented:
> I find it a little hard to review, it would be nice to split into smaller PRs for review.

I'd rather keep this in one PR as it's all related to the Directory API cleanup, splitting this up also generates extra considerations when updating moving parts, note that I already have https://gitlab.com/BuildStream/bst-plugins-experimental/-/merge_requests/166 lined up for this.

I had considered trying to reduce the number of commits (which would be quite difficult short of squashing it all into one commit) but I would like to keep them all in history, every commit here passes it's own tests and tells a full story of how things were changed here which is nice to keep.

[...]
> I have a couple of comments on these:
> 
>     * storage/directory.py: Remove `can_link` parameter from Directory.import_files()
>       Even though it is currently unused, there is potential to have a similar option that uses  the `move_files` hint  (by passing `--capture-allow-file-moves` to `buildbox-casd`). Can be done later though.

Sure I'll consider this unrelated, in any case it's highly unlikely that we would want to expose the abstract plugin API to such low level implementation details, given that the public API can never again be changed.

>     * storage/directory.py: Make Directory.export_files() an internal function
>       If we want to go for source.py: Split up staging functions into separate codepaths. #1607 (comment), 
>       we'll need to keep this public. Again, this can be made public later if needed (even if it's not at 2.0).

Why ? `Directory.export_files()` is only ever used by the core when *exporting* files in a virtual directory to the local filesystem, when e.g. checking out artifacts or sources, etc. `Directory.import_files()` of course needs to stay public.

juergbi commented on src/buildstream/storage/directory.py:
```diff
@@ -16,328 +17,499 @@
+Directory - Interfacing with files
+==================================
+The Directory class is given to plugins by way of the :class:`.Sandbox`
+or in some other instances, and allows plugins to interface with files
```
Word missing after 'or'?

gtristan replied:
I did `s/or/and/` here and improved the grammar of this phrase.

juergbi commented on src/buildstream/storage/directory.py:
```diff
@@ -16,328 +17,499 @@
+    #
+    def _validate_path(self, path: str) -> None:
+        if len(path) and path[0] == "/":
+            raise ValueError("Invalid path '{}'".format(path))
```
This function considers empty paths valid. Do we really want this or is this necessary? Intuitively I'd expect `.` to be accepted for the current directory but the empty path to be rejected, which matches POSIX behavior.

gtristan replied:
Ok let's do this.

gtristan replied:
Turns out this is not particularly user friendly, not sure if we should do this.

Elements like the `import` element want to call `inputdir = inputdir.descend(self.source.strip(os.sep))`, so that if there is a "/" then the input is at the root.

Forcing the caller to specify "." here seems to introduce more special casing on the caller side than is desirable.

Let's keep it this way.

juergbi commented on src/buildstream/storage/directory.py:
```diff
@@ -16,328 +17,499 @@
+:func:`Sandbox.get_virtual_directory() <buildstream.sandbox.Sandbox.get_virtual_directory>`
+is the root of the sandbox's *directory tree*. Attempts to escape the root of a *directory tree*
+using ``..`` entries will not result in an error, instead ``..`` entries which cross the
+root boundary will be evaluated as the root directory.
```
Maybe mention that this matches POSIX behavior of filesystem roots?

gtristan replied:
Appended to this paragraph:

> This behavior matches POSIX behavior of filesystem root directories.

juergbi commented on src/buildstream/storage/directory.py:
```diff
@@ -16,328 +17,499 @@
+    #                           Public API                            #
+    ###################################################################
+
+    def descend(self, path: str, *, create: bool = False, follow_symlinks: bool = False) -> "Directory":
```
As this also accepts .. (protected against escaping), the name may be a bit off. descend is also not common terminology when working with directories, as far as I know. Could it make sense to rename it to open_directory()? This would be consistent with open_file() and would be the Directory equivalent of POSIX open(..., O_DIRECTORY).

gtristan replied:
Done, let's call it `open_directory()` instead.

juergbi commented on src/buildstream/storage/directory.py:
```diff
@@ -16,328 +17,499 @@
     # Convenience functions
-    def is_empty(self):
-        """ Return true if this directory has no files, subdirectories or links in it.
+    def is_empty(self) -> bool:
```
The more Pythonic way would probably be to implement the inverse as `__bool__` (or `__len__` but that might not be needed).

gtristan replied:
Using `__len__` here since that seems to make sense because the `Directory` already otherwise behaves as a collection of entries.

Also updating the docs note about iterating over entries to mention that directories behave like collections of entries, and that they are falsy when emtpy, etc.

This was tricky as I needed to track down various instances where we need to change `if directory` to `if directory is not None`

juergbi commented on src/buildstream/storage/_casbaseddirectory.py:
```diff
@@ -114,136 +128,337 @@ def get_equivalency_properties(e: IndexEntry):
+    def is_empty(self) -> bool:
+        return len(self.__index) == 0
+
+    def listdir(self) -> Iterator[str]:
```
This is not declared in `Directory`. I'm wondering whether we should combine this and `list_relative_paths()` into a single method with an optional `recursive` (bool keyword-only) parameter. What do you think?

gtristan replied:
oops listdir() is an artifact of something I've already backed out of the branch.

This is handled by way of the `Directory` being iterable instead, I'll remove this from wherever it was added.

gtristan committed:
storage/directory.py: Remove is_empty() in favor of implementing __len__()

This is a more pythonic approach for the API.

gtristan committed:
storage/directory.py: Rename descend() -> open_directory()

Since we support ".." entries in this API this appears to be a better
name for the API and is consistent with the existing open_file() API name.
